src/services/jsTyping.ts

```diff
@@ -0,0 +1,246 @@
+
+/// <reference path='services.ts' />
+
+/* @internal */
+namespace ts.JsTyping {
+
+    interface TypingResolutionHost {
+        directoryExists: (path: string) => boolean;
+        fileExists: (fileName: string) => boolean;
+        readFile: (path: string, encoding?: string) => string;
+        readDirectory: (path: string, extension?: string, exclude?: string[], depth?: number) => string[];
+    };
+
+    interface TsdJson {
+        version: string;
+        repo: string;
+        ref: string;
+        path: string;
+        installed?: Map<TsdInstalledItem>;
+    };
+
+    interface TsdInstalledItem {
+        commit: string;
+    };
+
+    interface PackageJson {
+        _requiredBy?: string[];
+        dependencies?: Map<string>;
+        devDependencies?: Map<string>;
+        name: string;
+        optionalDependencies?: Map<string>;
+        peerDependencies?: Map<string>;
+        typings?: string;
+    };
+
+    // A map of loose file names to library names
+    // that we are confident require typings
+    let safeList: Map<string>;
+
+    /**
+     * @param host is the object providing I/O related operations.
+     * @param fileNames are the file names that belong to the same project.
+     * @param cachePath is the path to the typings cache
+     * @param projectRootPath is the path to the project root directory
+     * @param safeListPath is the path used to retrieve the safe list
+     * @param typingOptions are used for customizing the typing inference process.
+     * @param compilerOptions are used as a source of typing inference.
+     */
+    export function discoverTypings(
+        host: TypingResolutionHost,
+        fileNames: string[],
+        cachePath: Path,
+        projectRootPath: Path,
+        safeListPath: Path,
+        typingOptions: TypingOptions,
+        compilerOptions: CompilerOptions):
+        { cachedTypingPaths: string[], newTypingNames: string[], filesToWatch: string[] } {
+
+        // A typing name to typing file path mapping
+        const inferredTypings: Map<string> = {};
+
+        if (!typingOptions || !typingOptions.enableAutoDiscovery) {
+            return { cachedTypingPaths: [], newTypingNames: [], filesToWatch: [] };
+        }
+
+        // Only infer typings for .js and .jsx files
+        fileNames = filter(map(fileNames, normalizePath), f => scriptKindIs(f, /*LanguageServiceHost*/ undefined, ScriptKind.JS, ScriptKind.JSX));
+
+        if (!safeList) {
+            const result = readConfigFile(safeListPath, host.readFile);
+            if (result.config) { safeList = result.config; }
+        }
+
+        const filesToWatch: string[] = [];
+        // Directories to search for package.json, bower.json and other typing information
+        let searchDirs: string[] = [];
+        let exclude: string[] = [];
+
+        mergeTypings(typingOptions.include);
+        exclude = typingOptions.exclude || [];
+
+        if (typingOptions.enableAutoDiscovery) {

```

Comment:
I don't agree, because there are cases where the `include` array has content while the `enableAutoDiscovery` was set to `false`. This happens when the user knows exactly what he wants, therefore specifies all typings and doesn't want the inferring part.

The previous checking should not only checking `enableAutoDiscovery`. The original `isTypingEnabled` method should be kept as it considers other cases too.